drivers/LargeBlockSR.py

```diff
@@ -0,0 +1,201 @@
+            os.unlink(new_path)
+
+        try:
+            os.symlink(self.loopdev, new_path)
+        except OSError:
+            #util.pread2(["losetup", "-d", self.loopdev])
+            raise xs_errors.XenError("LargeBlockSymlinkExist", opterr="Symlink {} couldn't be created".format(new_path))
+
+    def _delete_loopdev(self, dev, new_path):
+            if os.path.exists(new_path) and os.path.islink(new_path):
+                os.unlink(new_path)
+
+            # The backing file isn't a symlink if given by ID in device-config but the real device
+            dev = os.path.realpath(dev)
+            loopdevs = self._get_loopdev_from_device(dev)
+
+            if loopdevs != None:
+                try:
+                    for lp in loopdevs:
+                        cmd = ["losetup", "-d", lp] # Remove the loop device
+                        util.pread2(cmd)
+                except SR.SROSError:
+                    util.SMlog("Couldn't removed losetup devices: {}".format(loopdevs))
+            else:
+                xs_errors.XenError("LargeBlockNoLosetup", opterr="Couldn't find loop device for {}".format(dev))
+
+    def _get_loopdev_from_device(self, device):
+        lpdevs = []
+        output = util.pread2(["losetup", "--list"]).rstrip()
+        if output:
+            for line in output.split("\n"):
+                line = line.split()
+                loopdev = line[0]
+                dev = line[5]
+                if device in dev:
+                    lpdevs.append(loopdev)
+        return lpdevs
+
+    def _is_vg_connection_correct(self):
+        output = util.pread2(["vgs", "--noheadings", "-o", "vg_name,devices", self.vgname]).split()
+        if self.vgname == output[0]:
+            output[1] = output[1].split("(")[0]
+        if re.match(r"(.*\.512)|(/dev/loop[0-9]+)", output[1]):
+            return True
+        else:
+            return False
+
+    def _redo_vg_connection(self):
+        """
+        In case of using a LargeBlockSR, the LVM scan at boot will find the LogicalVolume on the real block device.
+        And when the PBD is connecting, it will mount from the original device instead of the loop device since LVM prefers real devices it has seen first.
+        The PBD plug will succeed but then the SR will be accessed through the 4KiB device, returning to the erroneous behaviour on 4KiB device.
+        VM won't be able to run because vhd-util will fail to scan the VDI.
+        This function force the LogicalVolume to be mounted on top of our emulation layer by disabling the VolumeGroup and re-enabling while applying a filter.
+        """
+
+        util.SMlog("Reconnecting VG {} to use emulated device".format(self.vgname))
+        try:
+            cmd = ["vgchange", "-an", self.vgname]
+            util.pread2(cmd)
+            cmd = ["vgchange", "-ay", self.vgname, "--config", "'devices{ global_filter = [ \"r|^/dev/nvme.*|\", \"a|/dev/loop.*|\" ] }"]
+            util.pread2(cmd)
```

Comment:
I don't know why we don't use directly util.pread2 without and additional cmd variable.
It's the case in many drivers. :p

Reply:
Maybe better to use an explicit filter on the /dev/loop to use.